src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(null, value, null);
+        if (isEmpty()) {
+            first = node;
+        } else {
+            last.next = node;
+            node.prev = last;
+        }
+        last = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        verifyIndexIsValid(index);
+        Node<T> node = new Node<>(null, value, null);
+        if (index == 0) {
+            node.next = first;
+            first.prev = node;
+            first = node;
+            size++;
+            return;
+        }
+        Node<T> nextNode = getNode(index);
+        nextNode.prev.next = node;
+        node.prev = nextNode.prev;
+        nextNode.prev = node;
+        node.next = nextNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        verifyIndexIsValid(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        verifyIndexIsValid(index);
```

Comment:
code duplication as verifyIndexIsValid is called inside getNode(). Please fix in all places